### ISSUE TITLE
Undefined results when using filesystems cloned using "dd" command

### ISSUE THREAD

**Issue by HaleTom** [digint/btrbk#205]
```
$ btrbk --version
btrbk command line client, version 0.26.0

$ sudo btrbk archive /media/backup/btrbk/home /media/backup-alt/btrbk/home
ERROR: Source and target subvolumes are on the same btrfs filesystem!
```

But they are different filesystems:
```
$ df -P /media/backup/btrbk/home
Filesystem     1024-blocks      Used Available Capacity Mounted on
-                613371904 388935948 221659924      64% /media/backup/btrbk/home
$ df -P /media/backup-alt/btrbk/home
Filesystem     1024-blocks      Used Available Capacity Mounted on
-                613371904 378059340 231693556      63% /media/backup-alt/btrbk/home
```

With different UUIDs:

```
$ lsblk -f /dev/sdb
NAME               FSTYPE      LABEL                     UUID                                   MOUNTPOINT
sdb                                                                                             
├─sdb1             ntfs        Seagate Backup Plus Drive 01D3117F9DB28EC0                       
├─sdb2             LVM2_member                           Fe97xi-Iz34-iH0N-1s85-UZY5-tQ0A-hAyHgW 
│ ├─2TB-backup     btrfs       backup                    xxxxxxxx-a88e-4ebd-a04b-9459366832b7   /media/backup
│ └─2TB-backup_alt btrfs       backup-alt                xxxxxxxx-34ad-4dfd-b163-882fd1fdd2e5   /media/backup-alt
└─sdb3             LVM2_member                           MtLb3p-MUle-8fyk-fy6m-z99n-V9mi-4j5DYd 
  └─2TB-backup_alt btrfs       backup-alt                xxxxxxxx-34ad-4dfd-b163-882fd1fdd2e5   /media/backup-alt
```
(`xxxxxxxx` redacted)

Here is the `mount` output:
```
$ mount|grep backup
/dev/mapper/2TB-backup on /media/backup type btrfs (ro,nosuid,nodev,noexec,relatime,lazytime,compress=zlib,space_cache,subvolid=257,subvol=/@,user,x-systemd.idle-timeout=30)
/dev/mapper/2TB-backup_alt on /media/backup-alt type btrfs (rw,nosuid,nodev,noexec,relatime,lazytime,compress=zlib,space_cache,subvolid=257,subvol=/@,user=ravi)
```

I did `dd` copy `backup-alt` to `backup`, but afterwards used `btrfstune -u` to generate a new UUID. I do this because it is *much* faster than using `btrfs send`.

**Comment by digint** [comment]
> I did dd copy backup-alt to backup, but afterwards used btrfstune -u to generate a new UUID. I do this because it is much faster than using btrfs send.

Using `dd` to copy btrfs (and others also) filesystems is a **bad idea**. If you do this, make sure not to mount (or even enable) the two identical block devices. If you do, you run into the *really bad situation* where the kernel discovers two btrfs filesystems having **identical uuid**.

In your case, `backup-alt` on `sdb2` AND `sdb3`! This confuses the kernel (and btrfs-progs) and leads to undefined behavior.

> ERROR: Source and target subvolumes are on the same btrfs filesystem!

For all sources in `/media/backup/btrbk/home` (recursive), btrbk traverses the target filesystem tree (`/dev/mapper/2TB-backup_alt`) and matches the uuid of the source subvolume. On positive match, btrbk errors out with the message above.

https://github.com/digint/btrbk/blob/master/btrbk#L4627
https://github.com/digint/btrbk/blob/master/btrbk#L2162

**Comment by HaleTom** [comment]
I'm aware of the dangers of mounting a FS with another visible having the same UUID. (This did not occur, `btrfstune -u` was run immediately after `dd`)

LVM LVs are allowed to span VGs... this is what is happening with `backup-alt`. Note 2 segments:
```
$ sudo lvdisplay 2TB/backup_alt
  --- Logical volume ---
  LV Path                /dev/2TB/backup_alt
  LV Name                backup_alt
  VG Name                2TB
  LV UUID                kLOPPD-7Dj9-kCtM-7WX9-9qtM-EwR2-mhDpaS
  LV Write Access        read/write
  LV Creation host, time svelte, 2017-11-19 11:03:11 +0700
  LV Status              available
  # open                 0
  LV Size                584.96 GiB
  Current LE             149750
  Segments               2
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     256
  Block device           254:5
```

These 2 segments should not confuse the kernel. (Granted, it was poor planning on setting up my HDD storage though!)

**Comment by HaleTom** [comment]
Perhaps the UUID search is going to far "up" the tree, ie above the filesystem level and into the LVM / partition level?

**Comment by digint** [comment]
> Perhaps the UUID search is going to far "up" the tree, ie above the filesystem level and into the LVM / partition level?

The tree is built using the output of `btrfs subvolume list -a -c -u -q -R /path/to/btrfs_mountpoint`, so no, its not possible to go above btrfs filesystem level.

> If not, then which UUIDs are equal?

Too bad this information is not logged.

Please try patching btrbk, line 4628:
```
-ERROR "Source and target subvolumes are on the same btrfs filesystem!";
+ERROR "Source and target subvolumes are on the same btrfs filesystem! target=\"$droot->{PRINT}\", uuid=$vol->{node}{uuid}";
```

Does this help?

**Comment by HaleTom** [comment]
Here is what I see:
```
$ sudo ./btrbk archive /media/backup /media/backup-alt
ERROR: Source and target subvolumes are on the same btrfs filesystem! target="/media/backup-alt/btrbk/home", uuid=e520d3fb-9b92-3f40-90ba-58d25a466e29
```

This relates to subvolumes:
```
$ sudo btrfs sub list -u /media/backup | grep e520d3fb
ID 264 gen 555 top level 260 uuid e520d3fb-9b92-3f40-90ba-58d25a466e29 path btrbk/home/home.20171122T0000
$ sudo btrfs sub list -u /media/backup-alt | grep e520d3fb
ID 264 gen 555 top level 260 uuid e520d3fb-9b92-3f40-90ba-58d25a466e29 path btrbk/home/home.20171122T0000
```

These *subvolumes* indeed have the same UUID. [This](https://unix.stackexchange.com/a/247546/143394) shows that only FS UUIDs need to be unique, but subvol UUIDs may be dupliated.

The subvols are NOT on the same *filesystem*, as shown by `df`:
```
$ df -P /media/backup/btrbk/home/home.20171122T0000
Filesystem     1024-blocks      Used Available Capacity Mounted on
-                613371904 382630760 227973688      63% /media/backup/btrbk/home/home.20171122T0000
$ df -P /media/backup-alt/btrbk/home/home.20171122T0000
Filesystem     1024-blocks      Used Available Capacity Mounted on
-                613371904 384076700 226824100      63% /media/backup-alt/btrbk/home/home.20171122T0000
```
So, in this case, the target is NOT on the same filesystem, but it *does* already exist.

Perhaps the *filesystem* UUID could be checked in short-circuit logic evaluation, something like:

```
if( !_is_same_fs($droot->{node}->{TREE_ROOT}, $vol->{node}{uuid})) && _is_child_of($droot->{node}->{TREE_ROOT}, $vol->{node}{uuid})) {
  ERROR "Source and target subvolumes are on the same btrfs filesystem!";
  exit 1;
}
```

**Comment by HaleTom** [comment]
FYI, I commented out the `ERROR` and `exit 1` lines and added a warning.

I now get:
```
$ sudo ./btrbk archive --progress -vn /media/backup/btrbk/ /media/backup-alt/btrbk/
btrbk command line client, version 0.26.0  (Thu Jan 18 20:20:45 2018)
Using configuration: /etc/btrbk/btrbk.conf
WARNING: Skipping _is_child_of check
WARNING: Skipping _is_child_of check
WARNING: Skipping _is_child_of check
Archiving subvolumes: /media/backup/btrbk/home/home.*
No missing archives found
Archiving subvolumes: /media/backup/btrbk/home-var/home-var.*
No missing archives found
Archiving subvolumes: /media/backup/btrbk/rootfs/rootfs.*
No missing archives found
Cleaning archive: /media/backup-alt/btrbk/home/home.*
Deleted 0 subvolumes in: /media/backup-alt/btrbk/home/home.*
Cleaning archive: /media/backup-alt/btrbk/home-var/home-var.*
Deleted 0 subvolumes in: /media/backup-alt/btrbk/home-var/home-var.*
Cleaning archive: /media/backup-alt/btrbk/rootfs/rootfs.*
Deleted 0 subvolumes in: /media/backup-alt/btrbk/rootfs/rootfs.*
Completed within: 0s  (Thu Jan 18 20:20:45 2018)
```

Proving that there ARE missing backups:
```
$ diff <(ls /media/backup/btrbk/home) <(ls /media/backup-alt/btrbk/home) | tail -n5
> home.20171228T2100
> home.20171229T1956
8,9d43
< home.20180113T1854
< home.20180118T1421
```

**Comment by digint** [comment]
Well, I fear this will not be easy to fix. btrbk assumes that UUID's are *universally unique* and uses them as hash keys all over the place, e.g. `$uuid_cache` when building the subvolume tree in `btr_tree()`. This is probably why btrbk does not find your missing archives, as it ends up comparing the same (sub-)tree due to the cache.

Still btrbk should at least recognize such a scenario. There is already a check on [btrbk line 2065](/digint/btrbk/blob/v0.26.0/btrbk#L2065), but this only checks the root (filesystem) UUID. This might be enough for users who do a `dd` but don't `btrfstune -u` it.

@HaleTom could you try inserting the following code on [line 2078](/digint/btrbk/blob/v0.26.0/btrbk#L2078) and tell me if it bails out correctly:

```
$id{$node->{id}} = $node;
die "Duplicate subvolume UUID on different file systems: \"$node->{uuid}\"" if(exists($uuid_cache{$node->{uuid}}));
# ^^^ INSERT THIS LINE ^^^
$uuid_cache{$node->{uuid}} = $node;
```

**Comment by digint** [comment]
I tracked the problem further down to [line 2061](https://github.com/digint/btrbk/blob/v0.26.0/btrbk#L2061): btrbk identifies two filesystem trees as identical if the first (!) UUID listed in `btrfs subvolume list` is already known, and uses the cached tree instead of parsing it again.

@HaleTom If you disable this check (comment out the whole if() block at [line 2062](https://github.com/digint/btrbk/blob/v0.26.0/btrbk#L2062)), the target filesystem tree is reparsed and subsequent checks should not fail any more. Note that duplicate subvolume uuid's will still mess up the `$uuid_cache`, and btrbk might still break at other places.

Also note the new [FAQ entry](https://github.com/digint/btrbk/blob/master/doc/FAQ.md#warning-avoid-using-dd-on-btrfs-filesystems) on this (commit 308444f3ede8c5dee8d9c6ede4bac7fa063e4f1c).

**Comment by HaleTom** [comment]
Inserting:
```
die "Duplicate subvolume UUID on different file systems: \"$node->{uuid}\"" if(exists($uuid_cache{$node->{uuid}}));
```

Produced no additional output to the terminal.

Commenting out the `if` block produced no change in behaviour. 
(Note I also commented out the `die` at [line 4629](https://github.com/digint/btrbk/blob/v0.26.0/btrbk#L4629).)

```
$ sudo ~ravi/bin/btrbk archive -vn /media/backup /media/backup-alt
btrbk command line client, version 0.26.0  (Fri Jan 26 22:32:45 2018)
Using configuration: /etc/btrbk/btrbk.conf
WARNING: Source and target subvolumes are on the same btrfs filesystem! target="/media/backup-alt/btrbk/home", uuid=e520d3fb-9b92-3f40-90ba-58d25a466e29. Continuing.
WARNING: Source and target subvolumes are on the same btrfs filesystem! target="/media/backup-alt/btrbk/home-var", uuid=bd1e7f60-1390-5145-aa10-4f7f62da8a25. Continuing.
WARNING: Source and target subvolumes are on the same btrfs filesystem! target="/media/backup-alt/btrbk/rootfs", uuid=8abceb58-02ea-6947-8c20-c9bcf4360b63. Continuing.
Archiving subvolumes: /media/backup/btrbk/home/home.*
No missing archives found
Archiving subvolumes: /media/backup/btrbk/home-var/home-var.*
No missing archives found
Archiving subvolumes: /media/backup/btrbk/rootfs/rootfs.*
No missing archives found
Cleaning archive: /media/backup-alt/btrbk/home/home.*
Deleted 0 subvolumes in: /media/backup-alt/btrbk/home/home.*
Cleaning archive: /media/backup-alt/btrbk/home-var/home-var.*
Deleted 0 subvolumes in: /media/backup-alt/btrbk/home-var/home-var.*
Cleaning archive: /media/backup-alt/btrbk/rootfs/rootfs.*
Deleted 0 subvolumes in: /media/backup-alt/btrbk/rootfs/rootfs.*
Completed within: 0s  (Fri Jan 26 22:32:45 2018)
--------------------------------------------------------------------------------
Archive Summary (btrbk command line client, version 0.26.0)

    Date:   Fri Jan 26 22:32:45 2018
    Dryrun: YES

Legend:
    ++.  created directory
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------

NOTE: Dryrun was active, none of the operations above were actually executed!
$ 
```

**Comment by digint** [comment]
I expected something like this, seems like btrbk now gets confused with UUID's pointing to the wrong location, and recognizes them as already present (no more errors, just "No missing archives found").

btrbk uses subvolume UUID as unique identifiers as part of its hash-based subvolume tree management, so I'm afraid this is not reliably fixable without changing the whole concept.

resolved (wontfix) in: 308444f3ede8c5dee8d9c6ede4bac7fa063e4f1c, 753d68b2ed65d4aec0395fc92a54a7963777f77a

**Comment by HaleTom** [comment]
Thanks for your time on this.
